### ISSUE TITLE
Nesting PointerInterceptor platform views with Dropdown will cause crash on iOS

### ISSUE THREAD

**Issue by wenkaifan0720** [flutter/flutter#151535]
### Steps to reproduce

1. Create a `DropdownButton` wrapped in a `PointerInterceptor`.
2. Populate the items in the `DropdownButton` with some widget that is also wrapped in a `PointerInterceptor`.
3. Launch on iOS and click on the dropdown. The app will crash in a few seconds. 

### Expected results

The app doesn't crash.

### Actual results

The app crashes within a few seconds.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:pointer_interceptor/pointer_interceptor.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String? selected = "apple";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: PointerInterceptor(
            child: Padding(
              padding: MediaQuery.viewInsetsOf(context),
              child: Container(
                height: MediaQuery.sizeOf(context).height * 1,
                padding: EdgeInsets.all(40),
                child: DropdownButton<String>(
                  value: selected,
                  onChanged: (val) {
                    setState(() {
                      selected = val;
                    });
                  },
                  items: [
                    "apple",
                    "banana",
                    "cherry",
                    "dragonfruit",
                    "elderberry",
                    "fig",
                    "grape",
                    "honeydew",
                    "kiwi",
                    "lemon",
                    "mango",
                    "nectarine",
                    "orange",
                    "papaya",
                    "quince",
                  ]
                      .map(
                        (option) => DropdownMenuItem(
                      value: option,
                      child: PointerInterceptor(child: Text(option)),
                    ),
                  )
                      .toList(),
                ),
              ),
            )),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
Lost connection to device.
Error executing osascript: -2

the Dart compiler exited unexpectedly.
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.5 23F79 darwin-arm64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.3)
[✓] VS Code (version 1.91.0)
[✓] Connected device (4 available)
    ! Error: Browsing on the local area network for iPad (2). Ensure the device is unlocked and attached with a cable or
      associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
[✓] Network resources

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
Thanks for the report @wenkaifan0720 
I ran the code with latest stable in iOS simulator and was able to select items from dropdown. Also waited for sometime but app didn't crash. 
Are you running in simulator or physical device ? Please provide details of either.

**Comment by wenkaifan0720** [comment]
Hi @darshankawar. I tested on a real device (iPhone 13 mini, iOS 17.5.1).

**Comment by darshankawar** [comment]
I tried on iphone 6s but wasn't able to replicate the crash. The log says `The app “Runner” has been killed by the operating system because it is using too much memory.`, it could be due to combination of macos + xcode version.
You can use the memory profiler in flutter dev tools to find out how much memory is used. Also you could use Xcode Instruments to do a memory profile there to figure it out as what could be causing the memory leak.

**Comment by wenkaifan0720** [comment]
My Macos version is v14.5 (M1 pro Macbook Pro). My Xcode version is Version 15.4 (15F31d). I tried using the memory profiler, the memory usage is around 60MB and I don't see a spike before the crash. I am not very familiar with Xcode so I'll probably need more time to investigate on my own. The software I am using are all latest stable releases, so I do think this is a legitimate issue (and this issue has actually been reported for a few months by our users). And this issue will happen in both debug and release mode.

Below is a video recording of the crash:

https://github.com/user-attachments/assets/0b50db0d-e96b-45ea-ad51-4c6a4e710e42

It didn't happen the first time, but the crash is bound to happen if you interact more. This issue is affecting many users as this piece of code is being used by FlutterFlow. Could you try to test on a newer iPhone model?

**Comment by darshankawar** [comment]
Thanks for the update. Upon verifying on latest iOS version and config, we were able to replicate the crash as below:

```
The app "Runner" has been killed by the operating system because it is using too much memory.
Domain: IDEDebugSessionErrorDomain
Code: 11
Recovery Suggestion: Use a memory profiling tool to track the process memory usage.
User Info: {
    DVTErrorCreationDateKey = "2024-07-15 06:26:27 +0000";
    IDERunOperationFailingWorker = DBGLLDBLauncher;
}
--
The app "Runner" has been killed by the operating system because it is using too much memory.
Domain: IDEDebugSessionErrorDomain
Code: 11
Recovery Suggestion: Use a memory profiling tool to track the process memory usage.
User Info: {
    IDERunOperationFailingWorker = DBGLLDBLauncher;
}
--

Event Metadata: com.apple.dt.IDERunOperationWorkerFinished : {
    "device_isCoreDevice" = 1;
    "device_model" = "iPhone11,2";
    "device_osBuild" = "17.5.1 (21F90)";
    "device_platform" = "com.apple.platform.iphoneos";
    "dvt_coredevice_version" = "355.24";
    "dvt_mobiledevice_version" = "1643.100.58";
    "launchSession_schemeCommand" = Run;
    "launchSession_state" = 2;
    "launchSession_targetArch" = arm64;
    "operation_duration_ms" = 130695;
    "operation_errorCode" = 11;
    "operation_errorDomain" = IDEDebugSessionErrorDomain;
    "operation_errorWorker" = DBGLLDBLauncher;
    "operation_name" = IDERunOperationWorkerGroup;
    "param_debugger_attachToExtensions" = 0;
    "param_debugger_attachToXPC" = 1;
    "param_debugger_type" = 3;
    "param_destination_isProxy" = 0;
    "param_destination_platform" = "com.apple.platform.iphoneos";
    "param_diag_MainThreadChecker_stopOnIssue" = 0;
    "param_diag_MallocStackLogging_enableDuringAttach" = 0;
    "param_diag_MallocStackLogging_enableForXPC" = 1;
    "param_diag_allowLocationSimulation" = 1;
    "param_diag_checker_tpc_enable" = 1;
    "param_diag_gpu_frameCapture_enable" = 0;
    "param_diag_gpu_shaderValidation_enable" = 0;
    "param_diag_gpu_validation_enable" = 0;
    "param_diag_memoryGraphOnResourceException" = 0;
    "param_diag_queueDebugging_enable" = 1;
    "param_diag_runtimeProfile_generate" = 0;
    "param_diag_sanitizer_asan_enable" = 0;
    "param_diag_sanitizer_tsan_enable" = 0;
    "param_diag_sanitizer_tsan_stopOnIssue" = 0;
    "param_diag_sanitizer_ubsan_stopOnIssue" = 0;
    "param_diag_showNonLocalizedStrings" = 0;
    "param_diag_viewDebugging_enabled" = 1;
    "param_diag_viewDebugging_insertDylibOnLaunch" = 1;
    "param_install_style" = 0;
    "param_launcher_UID" = 2;
    "param_launcher_allowDeviceSensorReplayData" = 0;
    "param_launcher_kind" = 0;
    "param_launcher_style" = 99;
    "param_launcher_substyle" = 8192;
    "param_runnable_appExtensionHostRunMode" = 0;
    "param_runnable_productType" = "com.apple.product-type.application";
    "param_structuredConsoleMode" = 1;
    "param_testing_launchedForTesting" = 0;
    "param_testing_suppressSimulatorApp" = 0;
    "param_testing_usingCLI" = 0;
    "sdk_canonicalName" = "iphoneos17.4";
    "sdk_osVersion" = "17.4";
    "sdk_variant" = iphoneos;
}
--


System Information

macOS Version 14.4.1 (Build 23E224)
Xcode 15.3 (22618) (Build 15E204a)
Timestamp: 2024-07-15T06:26:27Z
```

<details>
<summary> stable, master flutter doctor -v </summary>

```

[✓] Flutter (Channel stable, 3.22.2, on macOS 14.4.1 23E224 darwin-arm64, locale en-GB)
    • Flutter version 3.22.2 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (6 weeks ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.0.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.91.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.92.0

[✓] Connected device (4 available)
    • Nexus (mobile)                  • 00008020-001875E83A38002E • ios            • iOS 17.5.1 21F90
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 126.0.6478.127
    ! Error: Browsing on the local area network for Dean's iPad. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

• No issues found!

  
[!] Flutter (Channel master, 3.24.0-1.0.pre.114, on macOS 14.4.1 23E224 darwin-arm64, locale en-GB)
    • Flutter version 3.24.0-1.0.pre.114 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 82b63ff27d (21 hours ago), 2024-07-14 05:07:36 -0400
    • Engine revision a921637f16
    • Dart version 3.6.0 (build 3.6.0-36.0.dev)
    • DevTools version 2.37.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.0.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.91.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.92.0

[✓] Connected device (4 available)
    • Nexus (mobile)                  • 00008020-001875E83A38002E • ios            • iOS 17.5.1 21F90
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 126.0.6478.127
    ! Error: Browsing on the local area network for Dean's iPad. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

```

</details>

**Comment by wenkaifan0720** [comment]
Thanks for confirming!

**Comment by mit-mit** [comment]
cc @jmagman looks like a crash regression

**Comment by jmagman** [comment]
I'm on main ToT 63b7fa5d62ccf42f3c352845286709ac00d7c53d, and I saw the large memory usage (hovering around 2-2.5GB) back to Flutter 3.16.6, though I usually wasn't able to get it to crash on my iPhone 14 (I don't have a iPhone 13 mini handy).

In Instruments, most of the memory is used by 60 QuartzCore `IOSurfaces` allocated from `FlutterPlatformViewsController`.
```	
CA::Render::create_iosurface_with_pixel_format(unsigned int, unsigned int, unsigned int, unsigned int, unsigned long long, __CFString const*)	
get_unused_drawable(_CAMetalLayerPrivate*, CAMetalLayerRotation, bool, bool)	
CAMetalLayerPrivateNextDrawableLocked(CAMetalLayer*, CAMetalDrawable**, unsigned long*)	
-[CAMetalLayer nextDrawable]	
flutter::GPUSurfaceMetalImpeller::AcquireFrame(SkISize const&)	
flutter::FlutterPlatformViewsController::SubmitFrame(GrDirectContext*, std::_fl::shared_ptr<flutter::IOSContext> const&, std::_fl::unique_ptr<flutter::SurfaceFrame, std::_fl::default_delete<flutter::SurfaceFrame>>)	
flutter::IOSExternalViewEmbedder::SubmitFlutterView(long long, GrDirectContext*, std::_fl::shared_ptr<impeller::AiksContext> const&, std::_fl::unique_ptr<flutter::SurfaceFrame, std::_fl::default_delete<flutter::SurfaceFrame>>)		
```

Each dropdown item is 7 native views. An example of a single dropdown item (like just "grape"): (from back to front):
```
<FlutterOverlayView: 0x13d590700; frame = (137.667 332; 22.3333 18); clipsToBounds = YES; autoresize = W+H; userInteractionEnabled = NO; layer = <CAMetalLayer: 0x30168ea00>>
<FlutterOverlayView: 0x13d590540; frame = (-137.667 -332; 390 844); autoresize = W+H; userInteractionEnabled = NO; layer = <CAMetalLayer: 0x301695020>>
<ChildClippingView: 0x13d3179c0; frame = (138.685 328; 20.6584 24); layer = <CALayer: 0x301885c00>>
<FlutterTouchInterceptingView: 0x13d3172c0; frame = (0 0; 20.6584 24); anchorPoint = (0, 0); gestureRecognizers = <NSArray: 0x301884d40>; layer = <CALayer: 0x301885600>>
<UIView: 0x13d317480; frame = (0 0; 20.3333 24); autoresize = W+H; backgroundColor = UIExtendedGrayColorSpace 0 0; layer = <CALayer: 0x301885ca0>>
<FlutterOverlayView: 0x13d590380; frame = (138.667 332; 20.6667 18); clipsToBounds = YES; autoresize = W+H; userInteractionEnabled = NO; layer = <CAMetalLayer: 0x30177ba80>>
<FlutterOverlayView: 0x13d5901c0; frame = (-138.667 -332; 390 844); autoresize = W+H; userInteractionEnabled = NO; layer = <CAMetalLayer: 0x30177b7e0>>
```

There are also some `FlutterOverlayView`s with a 1px height?
![Screenshot 2024-07-22 at 2 40 42 PM](https://github.com/user-attachments/assets/135916d3-fd1c-442b-8e6d-8ef9c9a19fa9)

I see 66 `FlutterOverlayView`s, 31 `FlutterTouchInterceptngView`s, and 31 `ChildClippingViews`.  There are 15 items in the list.

@hellohuanlin could you take a look at this?  Why are there so many `FlutterOverlayView`s?

**Comment by hellohuanlin** [comment]
It turns out that the platform view is added by the pointer interceptor here: https://github.com/flutter/packages/blob/2911b870677db466a41c8e44de8a96d2032dcfb7/packages/pointer_interceptor/pointer_interceptor_ios/lib/pointer_interceptor_ios.dart#L20

When we wrap a widget under `PointerInterceptor` widget, this code adds a platform view beneath it. The platform view needs an overlay view to display the widget on top of it. In your sample code, you used `PointerInterceptor` for both the menu, and every item under this menu. That's why we are seeing so many UIViews created. 

@wenkaifan0720 Can you try removing the inner touch interceptors and see if the outer one is sufficient? 

```
                      .map(
                        (option) => DropdownMenuItem(
                      value: option,
                      child: Text(option),
                    ),
                  )
```

My understanding is that as long as the item is on top of a platform view, the touch will be forwarded to the flutter widget above it. And the frame of the Menu widget itself should cover all the items in the menu, as highlighted here: 

![Screenshot 2024-07-23 at 10 14 38 AM](https://github.com/user-attachments/assets/e82536a7-a831-46c1-ade5-91097e7a5585)

**Comment by jmagman** [comment]
> That's why we are seeing so many UIViews created.

I believe the example is contrived to stress-test the scenario that there's a pointer interceptor in a widget interceptor wrapped in a widget with a pointer interceptor.  Given that, I'm wondering why every internal widget involves _7_ native views.

**Comment by wenkaifan0720** [comment]
@hellohuanlin I can do that. Since we auto generate these lines of code for our customers, it would mean adding some detection to prevent nesting `PointerInterceptor`s. However, if Flutter can fix this issue it would be even easier for us. I still think this is kind of an issue because there is no mentioning of not allowing nesting `PointerInterceptor` in the package's doc, and this code example does not crash on Android and web.

**Comment by hellohuanlin** [comment]
@wenkaifan0720 I'd like to confirm whether pointer interceptor still works with the above code. Thanks!

**Comment by wenkaifan0720** [comment]
Yes, it still works. In fact, removing both interceptors doesn't seem to affect our use case (webview at the bottom of a stack, opening a bottom sheet which contains a dropdown and interact with the dropdown).

**Comment by wenkaifan0720** [comment]
For our use case (a webview below, open a bottom sheet and then interact with a dropdown within the bottom sheet), on native iOS and Android things seem to be working even if I remove both `PointerInterceptor`. On web I still need all the `PointerInterceptor` for the app to work.

**Comment by hellohuanlin** [comment]
@wenkaifan0720 Thanks for confirming the code. 

> Given that, I'm wondering why every internal widget involves 7 native views.

These are the 7 views: 
- Child clipping view for clipping of platform view
- touch interceptor view, to relay the touch
- the platform view itself.
- 2 overlays per platform view, each overlay needs 2 UIViews: the overlay itself, and a wrapper for clipping of overlay. 

We may be able to merge touch interceptor and child clipping view, and get rid of the overlay wrapper, but i doubt it's gonna have any sizable performance improvement because they are quite lightweight. The heavier weight is the overlay, which involves the whole rendering pipeline (such as requesting a surface frame for the GPU to dump the pixels onto, which is a scarce system resource). We originally have a limit of 2 overlays per platform view, but with https://github.com/flutter/engine/pull/54010 landed, we now only create 1 overlay per platform view, so it should help a lot in this situation. 

> However, if Flutter can fix this issue it would be even easier for us. I still think this is kind of an issue because there is no mentioning of not allowing nesting PointerInterceptor in the package's doc

Nesting `PointerInterceptor` is allowed. It's just that it is not necessary as discussed in previous comments. The app crashed NOT because we nested `PointerInterceptor`, but rather due to the fact that we added so many `PointerInterceptor` (hence so many platform views), regardless of whether they are nested or not. 

Though our doc should have explained better that on iOS, platform views are expensive and there may be performance overhead by using this package. 

> and this code example does not crash on Android and web.

There's no android implementation of the package, so it should be a no-op on Android. Not sure about the web though, could be that web platform view isn't as expensive as iOS? (@yjbanov may know)

As a follow up I think I will update the doc, then we can close this.
